File: sdks/typescript/tests/unit/parse.test.ts
import { describe, it, expect, vi } from 'vitest';
import {
  extensionOf,
  parseTaskTypeFor,
  PARSE_PAGED_TASK_TYPES,
  PARSE_SUPPORTED_EXTENSIONS,
} from '../../src/parse/index.js';
import {
  createParse,
  type ConvertOptions,
  type ConvertRef,
  type ParseOptions,
  type ParseSource,
} from '../../src/parse-facade.js';

/** 建一对 parse / convert 门面，记录它们实际提交了什么任务、拿到了什么返回 */
const harness = (downResult: unknown) => {
  const created: { type: string; params?: Record<string, unknown>; [k: string]: unknown }[] = [];
  const createTask = vi.fn(async (params: never) => {
    created.push(params as never);
    return { id: 'task-1' } as never;
  });
  const { parse, convert } = createParse({
    createTask: createTask as never,
    waitTask: async (taskId) => ({ id: taskId, status: 'completed' }) as never,
    downTask: async () => downResult,
  });
  const run = (source: ParseSource, options?: ParseOptions) => parse(source, options);
  const runConvert = (ref: ConvertRef, options?: ConvertOptions) => convert(ref, options);
  return { run, runConvert, created };
};

/** parse 任务的最小合法返回：有 IR 引用才算数 */
const ir = (extra: Record<string, unknown> = {}) => ({
  irKey: '2026-08/ab/ir.json',
  irSchemaVersion: 'pptx.v1',
  ...extra,
});

describe('扩展名路由', () => {
  it('按扩展名映射到任务类型，大小写与查询串不影响判定', () => {
    expect(parseTaskTypeFor('a.pdf')).toBe('pdf.pdfParse');
    expect(parseTaskTypeFor('/tmp/deck.PPTX')).toBe('pptx.parse');
    expect(parseTaskTypeFor('report.docx')).toBe('docx.parseTextAndImage');
    expect(parseTaskTypeFor('https://x.com/a/b.key?v=1#f')).toBe('keynote.parseTextAndImage');
    expect(parseTaskTypeFor('a.txt')).toBeUndefined();
    expect(parseTaskTypeFor('noext')).toBeUndefined();
  });

  it('extensionOf 处理隐藏文件与无扩展名', () => {
    expect(extensionOf('.gitignore')).toBe('');
    expect(extensionOf('a/b/c.tar.gz')).toBe('.gz');
    expect(extensionOf('')).toBe('');
  });

  it('只有分页格式支持逐页 markdown', () => {
    expect([...PARSE_PAGED_TASK_TYPES].sort()).toEqual(['keynote.parseTextAndImage', 'pptx.parse']);
    expect(PARSE_SUPPORTED_EXTENSIONS).toEqual(['.pdf', '.pptx', '.docx', '.key']);
  });
});

describe('parse：只出 IR', () => {
  it('产出 IR 与它的引用，返回体原样透传', async () => {
    const raw = ir({ slides: [{}], slideMasters: [] });
    const { run, created } = harness(raw);
    const res = await run('./a.pptx');

    expect(created[0]).toMatchObject({ type: 'pptx.parse' });
    // 解析任务不再有任何 View 相关开关
    expect(created[0].params).toEqual({});
    expect(res).toEqual({
      taskId: 'task-1',
      type: 'pptx.parse',
      irKey: '2026-08/ab/ir.json',
      irSchemaVersion: 'pptx.v1',
      ir: raw,
    });
  });

  /**
   * 没有 irKey 说明服务端还没升到 parse/convert 拆分之后的版本：结果看着完整，却没有
   * 任何东西能交给 convert。在这里失败，比让调用方拿着 undefined 去调 convert 强。
   */
  it('返回体没有 irKey 时报错，指向服务端版本', async () => {
    const { run } = harness({ slides: [{}], slideMasters: [] });
    await expect(run('./a.pptx')).rejects.toThrow(/returned no irKey/);
    await expect(run('./a.pptx')).rejects.toThrow(/platform-slave 0\.22\.0/);
  });

  it('pdf 直通 password / parseProfile / includeImages', async () => {
    const { run, created } = harness(ir());
    await run('./a.pdf', { password: 'pw', parseProfile: 'quality', includeImages: false });

    expect(created[0].type).toBe('pdf.pdfParse');
    expect(created[0].params).toEqual({ password: 'pw', parseProfile: 'quality', includeImages: false });
  });

  it('keynote 直通 stayImageAreaRate', async () => {
    const { run, created } = harness(ir());
    await run('./a.key', { stayImageAreaRate: 0.08 });

    expect(created[0].type).toBe('keynote.parseTextAndImage');
    expect(created[0].params).toEqual({ stayImageAreaRate: 0.08 });
  });

  it('链接下发 url / mode，默认 runtime', async () => {
    const { run, created } = harness(ir());
    const res = await run({ url: 'https://example.com/a', mode: 'source' });

    expect(created[0]).toMatchObject({ type: 'html.getByURL' });
    expect(created[0].params).toEqual({ url: 'https://example.com/a', mode: 'source' });
    expect(res.type).toBe('html.getByURL');

    const second = harness(ir());
    await second.run({ url: 'https://example.com/a' });
    expect(second.created[0].params).toMatchObject({ mode: 'runtime' });
  });
});

describe('convert：IR → View', () => {
  const view = {
    format: 'pptx',
    schemaVersion: 'pptx.v1',
    to: 'markdown',
    markdown: '# hi',
    images: [],
  };

  it('按 irKey 转换，不重新上传源文件', async () => {
    const { runConvert, created } = harness(view);
    const res = await runConvert({ irKey: '2026-08/ab/ir.json' });

    expect(created[0]).toMatchObject({ type: 'parse.convert' });
    expect(created[0].params).toEqual({ irKey: '2026-08/ab/ir.json', to: 'markdown' });
    // 转换任务不带任何文件
    expect(created[0]).not.toHaveProperty('files');
    expect(created[0]).not.toHaveProperty('fileIds');
    expect(res).toEqual({ ...view, taskId: 'task-1' });
  });

  it('也可以直接引用产出该 IR 的 parse 任务', async () => {
    const { runConvert, created } = harness(view);
    await runConvert({ taskId: 'task-parse-1' });

    expect(created[0].params).toEqual({ taskId: 'task-parse-1', to: 'markdown' });
  });

  it('只下发显式传入的 View 开关，不塞默认值', async () => {
    const { runConvert, created } = harness(view);
    await runConvert({ irKey: 'k' }, { markdownPages: true, markdownMeta: false });

    expect(created[0].params).toEqual({ irKey: 'k', to: 'markdown', markdownPages: true, markdownMeta: false });
    expect(created[0].params).not.toHaveProperty('markdownStrict');
  });

  it('容错模式下透出 markdownError，不抛错', async () => {
    const { runConvert } = harness({ ...view, markdown: '', markdownError: 'boom' });
    await expect(runConvert({ irKey: 'k' })).resolves.toMatchObject({ markdown: '', markdownError: 'boom' });
  });

  /** 解析一次、反复转换：后续转换不应该再产生任何解析任务。 */
  it('同一份 IR 可以反复转换，源文件只解析一次', async () => {
    const parseHarness = harness(ir());
    const parsed = await parseHarness.run('./a.pptx');

    const convertHarness = harness(view);
    await convertHarness.runConvert({ irKey: parsed.irKey });
    await convertHarness.runConvert({ irKey: parsed.irKey, to: 'markdown' } as never);

    expect(parseHarness.created).toHaveLength(1);
    expect(convertHarness.created).toHaveLength(2);
    expect(convertHarness.created.every((task) => task.type === 'parse.convert')).toBe(true);
  });
});

describe('parse：输入形态', () => {
  it('fileId 输入靠 name 判定扩展名', async () => {
    const { run, created } = harness(ir());
    await run({ fileId: 'f-1', name: 'report.docx' });

    expect(created[0]).toMatchObject({ type: 'docx.parseTextAndImage', fileIds: ['f-1'] });
    expect(created[0]).not.toHaveProperty('files');
  });

  it('非字符串文件必须给 name，否则报错并列出支持的扩展名', async () => {
    const { run } = harness(ir());
    await expect(run({ file: new Uint8Array([1]) as never })).rejects.toThrow(
      /Supported extensions: \.pdf, \.pptx, \.docx, \.key/
    );
  });

  it('不支持的扩展名直接报错，不发任务', async () => {
    const { run, created } = harness(ir());
    await expect(run('./a.txt')).rejects.toThrow(/Cannot determine parser for "\.\/a\.txt"/);
    expect(created).toHaveLength(0);
  });
});

File: sdks/go/parse_test.go
package deckops

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"reflect"
	"strings"
	"testing"
	"time"
)

func TestExtensionRouting(t *testing.T) {
	for name, want := range map[string]TaskType{
		"a.pdf":                       TaskPDFParse,
		"/tmp/deck.PPTX":              TaskPptxParse,
		"report.docx":                 TaskDocxParse,
		"https://x.com/a/b.key?v=1#f": TaskKeynoteParse,
	} {
		got, ok := ParseTaskTypeFor(name)
		if !ok || got != want {
			t.Fatalf("ParseTaskTypeFor(%q) = %q, %v; want %q", name, got, ok, want)
		}
	}
	for _, name := range []string{"a.txt", "noext", ".gitignore"} {
		if got, ok := ParseTaskTypeFor(name); ok {
			t.Fatalf("ParseTaskTypeFor(%q) = %q; want no match", name, got)
		}
	}
	if ExtensionOf("a/b/c.tar.gz") != ".gz" || ExtensionOf("") != "" {
		t.Fatal("ExtensionOf mismatch")
	}
}

// parseServer serves the create → poll → download round trip and records the
// params each task was created with.
func parseServer(t *testing.T, taskType TaskType, download string, seen *[]map[string]any) *httptest.Server {
	t.Helper()
	return httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, request *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		switch {
		case request.Method == http.MethodPost && request.URL.Path == "/tools/tasks":
			var body map[string]any
			if err := json.NewDecoder(request.Body).Decode(&body); err != nil {
				t.Fatal(err)
			}
			*seen = append(*seen, body)
			if body["type"] != string(taskType) {
				t.Fatalf("task type = %#v, want %q", body["type"], taskType)
			}
			_, _ = w.Write([]byte(`{"id":"task-1","status":"pending"}`))
		case request.Method == http.MethodGet && request.URL.Path == "/tools/tasks/task-1":
			_, _ = w.Write([]byte(`{"id":"task-1","status":"completed"}`))
		case request.Method == http.MethodGet && request.URL.Path == "/tools/tasks/task-1/download":
			_, _ = w.Write([]byte(download))
		default:
			t.Fatalf("unexpected request: %s %s", request.Method, request.URL.Path)
		}
	}))
}

func testWait() WaitForTaskOptions {
	return WaitForTaskOptions{DisableSSE: true, Timeout: time.Second, PollInterval: time.Millisecond}
}

func newTestDeck(t *testing.T, ctx context.Context, root string) *Client {
	t.Helper()
	deck, err := New(ctx, ClientOptions{Root: root, SpaceID: "space-1", AuthUUID: testAuthUUID})
	if err != nil {
		t.Fatal(err)
	}
	return deck
}

func lastParams(t *testing.T, seen []map[string]any) map[string]any {
	t.Helper()
	if len(seen) == 0 {
		t.Fatal("no task was created")
	}
	params, _ := seen[len(seen)-1]["params"].(map[string]any)
	return params
}

func TestParseDefaultsToMarkdownOnly(t *testing.T) {
	resetAuthUUIDCacheForTests()
	ctx := context.Background()
	var seen []map[string]any
	server := parseServer(t, TaskPDFParse, `{"markdown":"# 标题","markdownImages":{"u":["k",1,"h"]}}`, &seen)
	defer server.Close()

	result, err := newTestDeck(t, ctx, server.URL).Parse(ctx,
		ParseSource{FileID: "file-1", Name: "report.PDF"},
		ParseOptions{Wait: testWait()},
	)
	if err != nil {
		t.Fatal(err)
	}
	if want := (map[string]any{"markdown": true, "markdownOnly": true}); !reflect.DeepEqual(lastParams(t, seen), want) {
		t.Fatalf("params = %#v, want %#v", lastParams(t, seen), want)
	}
	if result.Markdown != "# 标题" || result.TaskID != "task-1" || result.Type != TaskPDFParse {
		t.Fatalf("unexpected result: %#v", result)
	}
	if result.MarkdownImages["u"].Path != "k" {
		t.Fatalf("markdownImages = %#v", result.MarkdownImages)
	}
	if result.Result != nil {
		t.Fatalf("markdown output should not carry the raw payload: %s", result.Result)
	}
}

func TestParseOutputIRSkipsMarkdown(t *testing.T) {
	resetAuthUUIDCacheForTests()
	ctx := context.Background()
	var seen []map[string]any
	server := parseServer(t, TaskPDFParse, `{"document":{"elements":[]},"images":[]}`, &seen)
	defer server.Close()

	result, err := newTestDeck(t, ctx, server.URL).Parse(ctx,
		ParseSource{FileID: "file-1", Name: "a.pdf"},
		ParseOptions{Output: ParseOutputIR, Wait: testWait()},
	)
	if err != nil {
		t.Fatal(err)
	}
	if params := lastParams(t, seen); len(params) != 0 {
		t.Fatalf("ir output should not request markdown: %#v", params)
	}
	if result.Markdown != "" {
		t.Fatalf("ir output should not carry markdown: %q", result.Markdown)
	}
	var parsed PDFParseResult
	if err := json.Unmarshal(result.Result, &parsed); err != nil {
		t.Fatal(err)
	}
	if parsed.Document == nil {
		t.Fatalf("raw payload not passed through: %s", result.Result)
	}
}

func TestParseOutputAllPassesThroughParams(t *testing.T) {
	resetAuthUUIDCacheForTests()
	ctx := context.Background()
	var seen []map[string]any
	server := parseServer(t, TaskPptxParse, `{"slides":[{}],"markdown":"# a","markdownPages":["# a"]}`, &seen)
	defer server.Close()

	pages, strict := true, true
	result, err := newTestDeck(t, ctx, server.URL).Parse(ctx,
		ParseSource{FileID: "file-1", Name: "a.pptx"},
		ParseOptions{Output: ParseOutputAll, MarkdownPages: &pages, MarkdownStrict: &strict, Wait: testWait()},
	)
	if err != nil {
		t.Fatal(err)
	}
	want := map[string]any{"markdown": true, "markdownPages": true, "markdownStrict": true}
	if !reflect.DeepEqual(lastParams(t, seen), want) {
		t.Fatalf("params = %#v, want %#v", lastParams(t, seen), want)
	}
	if result.Markdown != "# a" || len(result.MarkdownPages) != 1 || result.Result == nil {
		t.Fatalf("unexpected result: %#v", result)
	}
}

func TestParsePDFPassthroughParams(t *testing.T) {
	resetAuthUUIDCacheForTests()
	ctx := context.Background()
	var seen []map[string]any
	server := parseServer(t, TaskPDFParse, `{"markdown":""}`, &seen)
	defer server.Close()

	images, meta := false, true
	_, err := newTestDeck(t, ctx, server.URL).Parse(ctx,
		ParseSource{FileID: "file-1", Name: "a.pdf"},
		ParseOptions{
			Password:      "pw",
			ParseProfile:  PDFParseProfileQuality,
			IncludeImages: &images,
			MarkdownMeta:  &meta,
			Wait:          testWait(),
		},
	)
	if err != nil {
		t.Fatal(err)
	}
	want := map[string]any{
		"markdown":      true,
		"markdownOnly":  true,
		"password":      "pw",
		"parseProfile":  "quality",
		"includeImages": false,
		"markdownMeta":  true,
	}
	if !reflect.DeepEqual(lastParams(t, seen), want) {
		t.Fatalf("params = %#v, want %#v", lastParams(t, seen), want)
	}
}

func TestParseDropsParamsTheTaskTypeRejects(t *testing.T) {
	resetAuthUUIDCacheForTests()
	ctx := context.Background()
	var seen []map[string]any
	server := parseServer(t, TaskDocxParse, `{"markdown":""}`, &seen)
	defer server.Close()

	pages := true
	rate := 0.08
	_, err := newTestDeck(t, ctx, server.URL).Parse(ctx,
		ParseSource{FileID: "file-1", Name: "a.docx"},
		ParseOptions{MarkdownPages: &pages, StayImageAreaRate: &rate, Wait: testWait()},
	)
	if err != nil {
		t.Fatal(err)
	}
	want := map[string]any{"markdown": true, "markdownOnly": true}
	if !reflect.DeepEqual(lastParams(t, seen), want) {
		t.Fatalf("params = %#v, want %#v", lastParams(t, seen), want)
	}
}

func TestParseKeynotePassthroughParams(t *testing.T) {
	resetAuthUUIDCacheForTests()
	ctx := context.Background()
	var seen []map[string]any
	server := parseServer(t, TaskKeynoteParse, `{"markdown":""}`, &seen)
	defer server.Close()

	pages := true
	rate := 0.08
	_, err := newTestDeck(t, ctx, server.URL).Parse(ctx,
		ParseSource{FileID: "file-1", Name: "a.key"},
		ParseOptions{MarkdownPages: &pages, StayImageAreaRate: &rate, Wait: testWait()},
	)
	if err != nil {
		t.Fatal(err)
	}
	want := map[string]any{
		"markdown":          true,
		"markdownOnly":      true,
		"markdownPages":     true,
		"stayImageAreaRate": 0.08,
	}
	if !reflect.DeepEqual(lastParams(t, seen), want) {
		t.Fatalf("params = %#v, want %#v", lastParams(t, seen), want)
	}
}

func TestParseURLSendsModeAndMarkdownSwitches(t *testing.T) {
	resetAuthUUIDCacheForTests()
	ctx := context.Background()
	var seen []map[string]any
	server := parseServer(t, TaskHTMLGetByURL, `{"markdown":"# page"}`, &seen)
	defer server.Close()

	result, err := newTestDeck(t, ctx, server.URL).Parse(ctx,
		ParseSource{URL: "https://example.com/posts/1"},
		ParseOptions{Wait: testWait()},
	)
	if err != nil {
		t.Fatal(err)
	}
	want := map[string]any{
		"url":          "https://example.com/posts/1",
		"mode":         string(ParseModeRuntime),
		"markdown":     true,
		"markdownOnly": true,
	}
	if !reflect.DeepEqual(lastParams(t, seen), want) {
		t.Fatalf("params = %#v, want %#v", lastParams(t, seen), want)
	}
	if result.Type != TaskHTMLGetByURL || result.Markdown != "# page" {
		t.Fatalf("unexpected result: %#v", result)
	}
}

func TestParseErrorsWhenBackendReturnsNoMarkdownField(t *testing.T) {
	// slave < 0.21.0 silently ignores the markdown params: the task succeeds but
	// the response carries no markdown key at all.
	resetAuthUUIDCacheForTests()
	ctx := context.Background()
	var seen []map[string]any
	server := parseServer(t, TaskPptxParse, `{"slides":[{}]}`, &seen)
	defer server.Close()

	_, err := newTestDeck(t, ctx, server.URL).Parse(ctx,
		ParseSource{FileID: "f", Name: "a.pptx"},
		ParseOptions{Wait: testWait()},
	)
	if err == nil || !strings.Contains(err.Error(), "returned no markdown field") {
		t.Fatalf("err = %v; want a missing-markdown error", err)
	}
}

func TestParseAcceptsPresentButEmptyMarkdown(t *testing.T) {
	resetAuthUUIDCacheForTests()
	ctx := context.Background()
	var seen []map[string]any
	server := parseServer(t, TaskPptxParse, `{"markdown":""}`, &seen)
	defer server.Close()

	result, err := newTestDeck(t, ctx, server.URL).Parse(ctx,
		ParseSource{FileID: "f", Name: "a.pptx"},
		ParseOptions{Wait: testWait()},
	)
	if err != nil {
		t.Fatalf("an empty document is a legitimate result, got %v", err)
	}
	if result.Markdown != "" {
		t.Fatalf("markdown = %q", result.Markdown)
	}
}

func TestParseOutputIRWorksAgainstOldBackend(t *testing.T) {
	resetAuthUUIDCacheForTests()
	ctx := context.Background()
	var seen []map[string]any
	server := parseServer(t, TaskPptxParse, `{"slides":[{}]}`, &seen)
	defer server.Close()

	result, err := newTestDeck(t, ctx, server.URL).Parse(ctx,
		ParseSource{FileID: "f", Name: "a.pptx"},
		ParseOptions{Output: ParseOutputIR, Wait: testWait()},
	)
	if err != nil {
		t.Fatalf("ir output must not depend on the markdown key: %v", err)
	}
	if result.Result == nil {
		t.Fatal("raw payload not passed through")
	}
}

func TestParseRejectsUnsupportedExtension(t *testing.T) {
	resetAuthUUIDCacheForTests()
	ctx := context.Background()
	deck := newTestDeck(t, ctx, "http://127.0.0.1:0")

	if _, err := deck.Parse(ctx, ParseSource{FileID: "f", Name: "a.txt"}, ParseOptions{}); err == nil {
		t.Fatal("expected an error for an unsupported extension")
	}
	if _, err := deck.Parse(ctx, ParseSource{FileID: "f", Name: "a.pdf"}, ParseOptions{Output: "bogus"}); err == nil {
		t.Fatal("expected an error for an unknown output tier")
	}
}
